prow/artifact-uploader/controller.go

```diff
@@ -0,0 +1,197 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package artifact_uploader
+
+import (
+	"bytes"
+	"fmt"
+	"path"
+	"time"
+
+	"github.com/golang/glog"
+	api "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/selection"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	core "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/test-infra/prow/pod-utils/decorate"
+
+	"k8s.io/test-infra/prow/gcsupload"
+	"k8s.io/test-infra/prow/kube"
+	"k8s.io/test-infra/prow/pod-utils/downwardapi"
+	"k8s.io/test-infra/prow/pod-utils/gcs"
+)
+
+const (
+	// ContainerLogDir is the prefix under which we place
+	// container logs in cloud storage
+	ContainerLogDir = "logs"
+)
+
+// item describes a container that we saw finish execution
+type item struct {
+	namespace     string
+	podName       string
+	containerName string
+	prowJobId     string
+}
+
+func NewController(client core.CoreV1Interface, prowJobClient *kube.Client, gcsConfig *gcsupload.Options) Controller {
+	queue := workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter())
+	optionsModifier := func(options *metav1.ListOptions) {
+		req, _ := labels.NewRequirement(kube.ProwJobIDLabel, selection.Exists, []string{})
+		options.LabelSelector = req.String()
+	}
+	podListWatcher := cache.NewFilteredListWatchFromClient(client.RESTClient(), "pods", api.NamespaceAll, optionsModifier)
+	indexer, informer := cache.NewIndexerInformer(podListWatcher, &api.Pod{}, 0, cache.ResourceEventHandlerFuncs{
+		UpdateFunc: func(old interface{}, new interface{}) {
+			oldPod := old.(*api.Pod)
+			newPod := new.(*api.Pod)
+
```

Comment:
Also do you want to act on the pod once it's marked for deletion? Elsewhere we avoid resyncing when newPod.DeletionTimestamp != nil.

Reply:
If it's marked for deletion and we gracefully terminate the containers we should still get logs for them, right? If we ignored the pod once it was marked for deletion would we lose those logs?

Reply:
For the most common case, sinker initiates the deletion and by that time you already have pushed the logs but you are probably right for voluntary evictions (node drains).